x-pack/libbeat/management/managerV2_test.go

```diff
@@ -0,0 +1,177 @@
+
+package management
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-client/v7/pkg/client/mock"
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+
+	"github.com/elastic/beats/v7/libbeat/common/reload"
+)
+
+func TestManagerV2(t *testing.T) {
+
+	srv := mockSrvWithUnits([]*proto.UnitExpected{
+		{
+			Id:             "output-unit",
+			Type:           proto.UnitType_OUTPUT,
+			ConfigStateIdx: 1,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "default",
+				Type: "elasticsearch",
+				Name: "elasticsearch",
+			},
+			State:    proto.State_HEALTHY,
+			LogLevel: proto.UnitLogLevel_INFO,
+		},
+		{
+			Id:             "input-unit-1",
+			Type:           proto.UnitType_INPUT,
+			ConfigStateIdx: 1,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "system/metrics-system-default-system-1",
+				Type: "system/metrics",
+				Name: "system-1",
+				Streams: []*proto.Stream{
+					{
+						Id: "system/metrics-system.filesystem-default-system-1",
+						Source: requireNewStruct(t, map[string]interface{}{
+							"metricsets": []interface{}{"filesystem"},
+							"period":     "1m",
+						}),
+					},
+				},
+			},
+			State:    proto.State_HEALTHY,
+			LogLevel: proto.UnitLogLevel_INFO,
+		},
+		{
+			Id:             "input-unit-2",
+			Type:           proto.UnitType_INPUT,
+			ConfigStateIdx: 1,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "system/metrics-system-default-system-2",
+				Type: "system/metrics",
+				Name: "system-2",
+				Streams: []*proto.Stream{
+					{
+						Id: "system/metrics-system.filesystem-default-system-2",
+						Source: requireNewStruct(t, map[string]interface{}{
+							"metricsets": []interface{}{"filesystem"},
+							"period":     "1m",
+						}),
+					},
+					{
+						Id: "system/metrics-system.filesystem-default-system-3",
+						Source: requireNewStruct(t, map[string]interface{}{
+							"metricsets": []interface{}{"filesystem"},
+							"period":     "1m",
+						}),
+					},
+				},
+			},
+			State:    proto.State_HEALTHY,
+			LogLevel: proto.UnitLogLevel_INFO,
+		},
+	})
+	require.NoError(t, srv.Start())
+	defer srv.Stop()
+
+	client := client.NewV2(fmt.Sprintf(":%d", srv.Port), "", client.VersionInfo{
+		Name:    "program",
+		Version: "v1.0.0",
+		Meta: map[string]string{
+			"key": "value",
+		},
+	}, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	r := reload.NewRegistry()
+
+	output := &reloadable{}
+	r.MustRegisterOutput(output)
+	inputs := &reloadableList{}
+	r.MustRegisterInput(inputs)
+
+	m, err := NewV2AgentManagerWithClient(&Config{
+		Enabled: true,
+	}, r, client)
+	require.NoError(t, err)
+
+	err = m.Start()
+	require.NoError(t, err)
+	defer m.Stop()
+
+	require.Eventually(t, func() bool {
```

Comment:
I think this would be a better test if we also transitioned some or all of the units expected states to stopped to ensure they are removed correctly.

It would also be nice to try to cover the logic to restart on output change but you'd have to mock out the actual stop/shutdown.

Reply:
I added more to the test to check for, state changes, log level changes, and stopping and removal of units.